File: Algorithm/programmers/pg_1829.py
from collections import deque

dr = (1,0),(0,1),(-1,0),(0,-1)

def Solution(m,n,picture):
    answer = [0,0]
    numberOfArea = 0
    maxSizeOfOneArea = 0
    visit = [[0]*n for _ in range(m)]

    for i in range(m):
        for j in range(n):
            if picture[i][j] !=0 and not visit[i][j]:
                target = picture[i][j]
                q = deque([(i,j)])
                visit[i][j] = 1
                cnt = 1
                while q:
                    x,y = q.popleft()

                    for dx,dy in dr:
                        di,dj = x+dx, y+dy
                        if 0<=di<m and 0<=dj<n and not visit[di][dj] and picture[di][dj]==target:
                            cnt += 1
                            visit[di][dj] = 1
                            q.append((di,dj))
                maxSizeOfOneArea = max(maxSizeOfOneArea, cnt)
                numberOfArea+=1

    answer[0] = numberOfArea
    answer[1] = maxSizeOfOneArea
    return answer

m=6
n=4
picture =[[1, 1, 1, 0], [1, 2, 2, 0], [1, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 3], [0, 0, 0, 3]]

print(Solution(m,n,picture))

File: Algorithm/baekjoon/java/src/pg_1829.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    static int[][] dr = {{1,0}, {0,1}, {-1,0}, {0,-1}};
    public int[] solution(int m, int n, int[][] picture) {
        int numberOfArea = 0;
        int maxSizeOfOneArea = 0;

        boolean[][] visit = new boolean[m][n];

        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                if (picture[i][j] !=0 && !visit[i][j]) {
                    int target = picture[i][j];
                    Queue<int[]> q = new LinkedList<>();
                    q.offer(new int[]{i,j});
                    visit[i][j] = true;
                    int cnt = 1;

                    while (!q.isEmpty()) {
                        int[] cur = q.poll();

                        int x = cur[0];
                        int y = cur[1];

                        for (int[] d : dr) {
                            int di = x + d[0];
                            int dj = y + d[1];
                            if ( 0<= di && di <m && 0<=dj && dj<n && !visit[di][dj] && picture[di][dj] == target) {
                                cnt ++;
                                visit[di][dj] = true;
                                q.offer(new int[]{di,dj});
                            }
                        }
                    }
                    maxSizeOfOneArea = Math.max(maxSizeOfOneArea,cnt);
                    numberOfArea ++ ;
                }
            }
        }


        int[] answer = new int[2];
        answer[0] = numberOfArea;
        answer[1] = maxSizeOfOneArea;
        return answer;
    }
}